### PR TITLE
Fix plugin classloading

### PR DESCRIPTION
**Pull request by robbavey**
Prior to this commit, attempting to use a custom java plugin installed from rubygems would
fail as the `files` field used to find jar files in the gem is not populated for such gems

This commit uses the `full_require_path` to find the jar file for a native jar plugin,
which should work regardless of where the gem was installed from, locally or via rubygems

Additionally, this commit will only attempt to locate the jar file inside the gem when it
is required (only when the experimental separate plugin class loader is used)

This commit builds on the work done in https://github.com/elastic/logstash/pull/13888

Relates: https://github.com/elastic/logstash/pull/13888, https://github.com/elastic/logstash/issues/11305

## Release notes
- Fixes an issue where custom java plugins were unable to be installed and run correctly when retrieved from rubygems.org

## Why is it important/What is the impact to the user?

This PR fixes an issue where custom java plugins from rubygems.org were not able to be installed and used

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files (and/or docker env variables)
- [x] I have added tests that prove my fix is effective or that my feature works

## How to test this PR locally

* This requires testing with a custom java plugin published to rubygems, plus a locally installed one
* For my testing of a published gem, I used:

```
./bin/logstash-plugin install logstash-filter-my_java_filter_example        
./bin/logstash -e "input { java_generator { message => 'Hello world!' count => 1 } } filter { my_java_filter_example {} } output { stdout { codec => rubydebug } }" 
```

* For the testing of a local Gem, I pulled down https://github.com/logstash-plugins/logstash-filter-java_filter_example and built locally, and then installed and ran:
```
./bin/logstash-plugin install --no-verify --local ../plugins/logstash-filter-java_filter_example/logstash-filter-java_filter_example-1.0.1.gem
./bin/logstash -e "input { java_generator { message => 'Hello world!' count => 1 } } filter { java_filter_example {} } output { stdout { codec => rubydebug } }" 
``` 

## Related issues

Closes #13888, #11305

### CODE REVIEW

File: logstash-core/src/main/java/org/logstash/plugins/PluginClassLoader.java [hunk 1]
```diff
@@ -44,18 +44,15 @@ private PluginClassLoader(URL[] urls, ClassLoader appClassLoader) {
 
     /**
      * Creates an instance of the plugin classloader.
-     * @param gemPath Path to the Ruby gem containing the Java plugin as reported by
-     *                <code>Gem::BasicSpecification#loaded_from</code>.
      * @param jarPath Path to the Java plugin's JAR file relative to {@code gemPath}.
```
Comment (yaauie) [inline logstash-core/src/main/java/org/logstash/plugins/PluginClassLoader.java:47]: nit: javadoc references removed `gemPath` parameter.

File: logstash-core/lib/logstash/plugins/registry.rb [hunk 2]
```diff
@@ -151,13 +151,14 @@ def load_available_plugins
       require "logstash/plugins/builtin"
 
       GemRegistry.logstash_plugins.each do |plugin_context|
-        if plugin_context.spec.metadata.key?('java_plugin')
-          jar_files = plugin_context.spec.files.select {|f| f =~ /.*\.jar/}
+        if plugin_context.spec.metadata.key?('java_plugin') && LogStash::SETTINGS.get_value('pipeline.plugin_classloaders')
+          # Find *.jar file from the require path
+          jar_files = Dir.glob(plugin_context.spec.full_require_paths.map{|path| "#{path}/**/*.jar"})
```
Comment (yaauie) [inline logstash-core/lib/logstash/plugins/registry.rb:156]: `Gem::Specification#matches_for_glob` is based on the require path and simplifies this a bit :)

```suggestion
          jar_files = plugin_context.spec.matches_for_glob("**/*.jar")
```

File: logstash-core/lib/logstash/plugins/registry.rb [hunk 3]
```diff
@@ -151,13 +151,14 @@ def load_available_plugins
       require "logstash/plugins/builtin"
 
       GemRegistry.logstash_plugins.each do |plugin_context|
-        if plugin_context.spec.metadata.key?('java_plugin')
-          jar_files = plugin_context.spec.files.select {|f| f =~ /.*\.jar/}
+        if plugin_context.spec.metadata.key?('java_plugin') && LogStash::SETTINGS.get_value('pipeline.plugin_classloaders')
```
Comment (yaauie) [inline logstash-core/lib/logstash/plugins/registry.rb:154]: While this bit of code effectively populates an entry in `@java_plugins` for use when `pipeline.plugin_classloaders: true`, is there still value in still performing the validation when `pipeline.plugin_classloaders: false`? Or are we relying on a runtime exception when the plugin attempts to require a jar and it is not found on the require path?

My worry here is that effectively removing this validation from the default configuration will result in users building plugins that work _by accident_, but fail when Logstash is configured with plugin classloaders.

**Review by yaauie — COMMENTED** [review]
I've left a nitpick about javadoc, a minor suggestion that doesn't change functionality, and a question about the value of removing validation of the shape of the plugin when the plugin classloaders are not enabled.

File: logstash-core/lib/logstash/plugins/registry.rb (hunk 2 above)
Comment (robbavey) [reply logstash-core/lib/logstash/plugins/registry.rb:156]: Gah, completely missed that method. Nice catch

File: logstash-core/lib/logstash/plugins/registry.rb (hunk 3 above)
Comment (robbavey) [reply logstash-core/lib/logstash/plugins/registry.rb:154]: That's fair. My intent was to allow the default (non-experimental) path to be as simple as possible, but I think you're right - removing the check opens the possibility of difficult to diagnose issues if a java plugin has been built and not packaged correctly.

**Comment by robbavey** [comment]
Ready for another round @yaauie

**Review by yaauie — APPROVED** [review]
LGTM 🍕

**Comment by robbavey** [comment]
@logstashmachine backport 8.2

**Pull request merged**